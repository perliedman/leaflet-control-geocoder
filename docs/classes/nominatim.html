<!doctype html>
<html class="default no-js">
<head>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<title>Nominatim | leaflet-control-geocoder</title>
	<meta name="description" content="Documentation for leaflet-control-geocoder">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="stylesheet" href="../assets/css/main.css">
</head>
<body>
<header>
	<div class="tsd-page-toolbar">
		<div class="container">
			<div class="table-wrap">
				<div class="table-cell" id="tsd-search" data-index="../assets/js/search.json" data-base="..">
					<div class="field">
						<label for="tsd-search-field" class="tsd-widget search no-caption">Search</label>
						<input id="tsd-search-field" type="text" />
					</div>
					<ul class="results">
						<li class="state loading">Preparing search index...</li>
						<li class="state failure">The search index is not available</li>
					</ul>
					<a href="../index.html" class="title">leaflet-control-geocoder</a>
				</div>
				<div class="table-cell" id="tsd-widgets">
					<div id="tsd-filter">
						<a href="#" class="tsd-widget options no-caption" data-toggle="options">Options</a>
						<div class="tsd-filter-group">
							<div class="tsd-select" id="tsd-filter-visibility">
								<span class="tsd-select-label">All</span>
								<ul class="tsd-select-list">
									<li data-value="public">Public</li>
									<li data-value="protected">Public/Protected</li>
									<li data-value="private" class="selected">All</li>
								</ul>
							</div>
							<input type="checkbox" id="tsd-filter-inherited" checked />
							<label class="tsd-widget" for="tsd-filter-inherited">Inherited</label>
							<input type="checkbox" id="tsd-filter-externals" checked />
							<label class="tsd-widget" for="tsd-filter-externals">Externals</label>
							<input type="checkbox" id="tsd-filter-only-exported" />
							<label class="tsd-widget" for="tsd-filter-only-exported">Only exported</label>
						</div>
					</div>
					<a href="#" class="tsd-widget menu no-caption" data-toggle="menu">Menu</a>
				</div>
			</div>
		</div>
	</div>
	<div class="tsd-page-title">
		<div class="container">
			<ul class="tsd-breadcrumb">
				<li>
					<a href="../globals.html">Globals</a>
				</li>
				<li>
					<a href="nominatim.html">Nominatim</a>
				</li>
			</ul>
			<h1>Class Nominatim</h1>
		</div>
	</div>
</header>
<div class="container container-main">
	<div class="row">
		<div class="col-8 col-content">
			<section class="tsd-panel tsd-comment">
				<div class="tsd-comment tsd-typography">
					<div class="lead">
						<p>Implementation of the <a href="https://wiki.openstreetmap.org/wiki/Nominatim">Nominatim</a> geocoder.</p>
					</div>
					<p>This is the default geocoding service used by the control, unless otherwise specified in the options.</p>
					<p>Unless using your own Nominatim installation, please refer to the <a href="https://operations.osmfoundation.org/policies/nominatim/">Nominatim usage policy</a>.</p>
				</div>
			</section>
			<section class="tsd-panel tsd-hierarchy">
				<h3>Hierarchy</h3>
				<ul class="tsd-hierarchy">
					<li>
						<span class="target">Nominatim</span>
					</li>
				</ul>
			</section>
			<section class="tsd-panel">
				<h3>Implements</h3>
				<ul class="tsd-hierarchy">
					<li><a href="../interfaces/igeocoder.html" class="tsd-signature-type">IGeocoder</a></li>
				</ul>
			</section>
			<section class="tsd-panel-group tsd-index-group">
				<h2>Index</h2>
				<section class="tsd-panel tsd-index-panel">
					<div class="tsd-index-content">
						<section class="tsd-index-section ">
							<h3>Constructors</h3>
							<ul class="tsd-index-list">
								<li class="tsd-kind-constructor tsd-parent-kind-class"><a href="nominatim.html#constructor" class="tsd-kind-icon">constructor</a></li>
							</ul>
						</section>
						<section class="tsd-index-section ">
							<h3>Methods</h3>
							<ul class="tsd-index-list">
								<li class="tsd-kind-method tsd-parent-kind-class"><a href="nominatim.html#geocode" class="tsd-kind-icon">geocode</a></li>
								<li class="tsd-kind-method tsd-parent-kind-class"><a href="nominatim.html#reverse" class="tsd-kind-icon">reverse</a></li>
							</ul>
						</section>
						<section class="tsd-index-section ">
							<h3>Object literals</h3>
							<ul class="tsd-index-list">
								<li class="tsd-kind-object-literal tsd-parent-kind-class"><a href="nominatim.html#options" class="tsd-kind-icon">options</a></li>
							</ul>
						</section>
					</div>
				</section>
			</section>
			<section class="tsd-panel-group tsd-member-group ">
				<h2>Constructors</h2>
				<section class="tsd-panel tsd-member tsd-kind-constructor tsd-parent-kind-class">
					<a name="constructor" class="tsd-anchor"></a>
					<h3>constructor</h3>
					<ul class="tsd-signatures tsd-kind-constructor tsd-parent-kind-class">
						<li class="tsd-signature tsd-kind-icon">new <wbr>Nominatim<span class="tsd-signature-symbol">(</span>options<span class="tsd-signature-symbol">?: </span><span class="tsd-signature-type">Partial</span><span class="tsd-signature-symbol">&lt;</span><a href="../interfaces/nominatimoptions.html" class="tsd-signature-type">NominatimOptions</a><span class="tsd-signature-symbol">&gt;</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><a href="nominatim.html" class="tsd-signature-type">Nominatim</a></li>
					</ul>
					<ul class="tsd-descriptions">
						<li class="tsd-description">
							<aside class="tsd-sources">
								<ul>
									<li>Defined in <a href="https://github.com/perliedman/leaflet-control-geocoder/blob/44cd093/src/geocoders/nominatim.ts#L89">src/geocoders/nominatim.ts:89</a></li>
								</ul>
							</aside>
							<h4 class="tsd-parameters-title">Parameters</h4>
							<ul class="tsd-parameters">
								<li>
									<h5><span class="tsd-flag ts-flagOptional">Optional</span> options: <span class="tsd-signature-type">Partial</span><span class="tsd-signature-symbol">&lt;</span><a href="../interfaces/nominatimoptions.html" class="tsd-signature-type">NominatimOptions</a><span class="tsd-signature-symbol">&gt;</span></h5>
								</li>
							</ul>
							<h4 class="tsd-returns-title">Returns <a href="nominatim.html" class="tsd-signature-type">Nominatim</a></h4>
						</li>
					</ul>
				</section>
			</section>
			<section class="tsd-panel-group tsd-member-group ">
				<h2>Methods</h2>
				<section class="tsd-panel tsd-member tsd-kind-method tsd-parent-kind-class">
					<a name="geocode" class="tsd-anchor"></a>
					<h3>geocode</h3>
					<ul class="tsd-signatures tsd-kind-method tsd-parent-kind-class">
						<li class="tsd-signature tsd-kind-icon">geocode<span class="tsd-signature-symbol">(</span>query<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">string</span>, cb<span class="tsd-signature-symbol">: </span><a href="../globals.html#geocodingcallback" class="tsd-signature-type">GeocodingCallback</a>, context<span class="tsd-signature-symbol">?: </span><span class="tsd-signature-type">any</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">void</span></li>
					</ul>
					<ul class="tsd-descriptions">
						<li class="tsd-description">
							<aside class="tsd-sources">
								<p>Implementation of <a href="../interfaces/igeocoder.html">IGeocoder</a>.<a href="../interfaces/igeocoder.html#geocode">geocode</a></p>
								<ul>
									<li>Defined in <a href="https://github.com/perliedman/leaflet-control-geocoder/blob/44cd093/src/geocoders/nominatim.ts#L95">src/geocoders/nominatim.ts:95</a></li>
								</ul>
							</aside>
							<h4 class="tsd-parameters-title">Parameters</h4>
							<ul class="tsd-parameters">
								<li>
									<h5>query: <span class="tsd-signature-type">string</span></h5>
								</li>
								<li>
									<h5>cb: <a href="../globals.html#geocodingcallback" class="tsd-signature-type">GeocodingCallback</a></h5>
								</li>
								<li>
									<h5><span class="tsd-flag ts-flagOptional">Optional</span> context: <span class="tsd-signature-type">any</span></h5>
								</li>
							</ul>
							<h4 class="tsd-returns-title">Returns <span class="tsd-signature-type">void</span></h4>
						</li>
					</ul>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-method tsd-parent-kind-class">
					<a name="reverse" class="tsd-anchor"></a>
					<h3>reverse</h3>
					<ul class="tsd-signatures tsd-kind-method tsd-parent-kind-class">
						<li class="tsd-signature tsd-kind-icon">reverse<span class="tsd-signature-symbol">(</span>location<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">LatLngLiteral</span>, scale<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">number</span>, cb<span class="tsd-signature-symbol">: </span><a href="../globals.html#geocodingcallback" class="tsd-signature-type">GeocodingCallback</a>, context<span class="tsd-signature-symbol">?: </span><span class="tsd-signature-type">any</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">void</span></li>
					</ul>
					<ul class="tsd-descriptions">
						<li class="tsd-description">
							<aside class="tsd-sources">
								<p>Implementation of <a href="../interfaces/igeocoder.html">IGeocoder</a>.<a href="../interfaces/igeocoder.html#reverse">reverse</a></p>
								<ul>
									<li>Defined in <a href="https://github.com/perliedman/leaflet-control-geocoder/blob/44cd093/src/geocoders/nominatim.ts#L120">src/geocoders/nominatim.ts:120</a></li>
								</ul>
							</aside>
							<h4 class="tsd-parameters-title">Parameters</h4>
							<ul class="tsd-parameters">
								<li>
									<h5>location: <span class="tsd-signature-type">LatLngLiteral</span></h5>
								</li>
								<li>
									<h5>scale: <span class="tsd-signature-type">number</span></h5>
								</li>
								<li>
									<h5>cb: <a href="../globals.html#geocodingcallback" class="tsd-signature-type">GeocodingCallback</a></h5>
								</li>
								<li>
									<h5><span class="tsd-flag ts-flagOptional">Optional</span> context: <span class="tsd-signature-type">any</span></h5>
								</li>
							</ul>
							<h4 class="tsd-returns-title">Returns <span class="tsd-signature-type">void</span></h4>
						</li>
					</ul>
				</section>
			</section>
			<section class="tsd-panel-group tsd-member-group ">
				<h2>Object literals</h2>
				<section class="tsd-panel tsd-member tsd-kind-object-literal tsd-parent-kind-class">
					<a name="options" class="tsd-anchor"></a>
					<h3>options</h3>
					<div class="tsd-signature tsd-kind-icon">options<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">object</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/perliedman/leaflet-control-geocoder/blob/44cd093/src/geocoders/nominatim.ts#L65">src/geocoders/nominatim.ts:65</a></li>
						</ul>
					</aside>
					<section class="tsd-panel tsd-member tsd-kind-variable tsd-parent-kind-object-literal">
						<a name="options.serviceurl" class="tsd-anchor"></a>
						<h3>service<wbr>Url</h3>
						<div class="tsd-signature tsd-kind-icon">service<wbr>Url<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">string</span><span class="tsd-signature-symbol"> = &quot;https://nominatim.openstreetmap.org/&quot;</span></div>
						<aside class="tsd-sources">
							<ul>
								<li>Defined in <a href="https://github.com/perliedman/leaflet-control-geocoder/blob/44cd093/src/geocoders/nominatim.ts#L66">src/geocoders/nominatim.ts:66</a></li>
							</ul>
						</aside>
					</section>
					<section class="tsd-panel tsd-member tsd-kind-function tsd-parent-kind-object-literal">
						<a name="options.htmltemplate" class="tsd-anchor"></a>
						<h3>html<wbr>Template</h3>
						<ul class="tsd-signatures tsd-kind-function tsd-parent-kind-object-literal">
							<li class="tsd-signature tsd-kind-icon">html<wbr>Template<span class="tsd-signature-symbol">(</span>r<span class="tsd-signature-symbol">: </span><a href="../interfaces/nominatimresult.html" class="tsd-signature-type">NominatimResult</a><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">string</span></li>
						</ul>
						<ul class="tsd-descriptions">
							<li class="tsd-description">
								<aside class="tsd-sources">
									<ul>
										<li>Defined in <a href="https://github.com/perliedman/leaflet-control-geocoder/blob/44cd093/src/geocoders/nominatim.ts#L67">src/geocoders/nominatim.ts:67</a></li>
									</ul>
								</aside>
								<h4 class="tsd-parameters-title">Parameters</h4>
								<ul class="tsd-parameters">
									<li>
										<h5>r: <a href="../interfaces/nominatimresult.html" class="tsd-signature-type">NominatimResult</a></h5>
									</li>
								</ul>
								<h4 class="tsd-returns-title">Returns <span class="tsd-signature-type">string</span></h4>
							</li>
						</ul>
					</section>
				</section>
			</section>
		</div>
		<div class="col-4 col-menu menu-sticky-wrap menu-highlight">
			<nav class="tsd-navigation primary">
				<ul>
					<li class="globals  ">
						<a href="../globals.html"><em>Globals</em></a>
					</li>
				</ul>
			</nav>
			<nav class="tsd-navigation secondary menu-sticky">
				<ul class="before-current">
				</ul>
				<ul class="current">
					<li class="current tsd-kind-class">
						<a href="nominatim.html" class="tsd-kind-icon">Nominatim</a>
						<ul>
							<li class=" tsd-kind-constructor tsd-parent-kind-class">
								<a href="nominatim.html#constructor" class="tsd-kind-icon">constructor</a>
							</li>
							<li class=" tsd-kind-method tsd-parent-kind-class">
								<a href="nominatim.html#geocode" class="tsd-kind-icon">geocode</a>
							</li>
							<li class=" tsd-kind-method tsd-parent-kind-class">
								<a href="nominatim.html#reverse" class="tsd-kind-icon">reverse</a>
							</li>
							<li class=" tsd-kind-object-literal tsd-parent-kind-class">
								<a href="nominatim.html#options" class="tsd-kind-icon">options</a>
							</li>
						</ul>
					</li>
				</ul>
				<ul class="after-current">
				</ul>
			</nav>
		</div>
	</div>
</div>
<footer class="with-border-bottom">
	<div class="container">
		<h2>Legend</h2>
		<div class="tsd-legend-group">
			<ul class="tsd-legend">
				<li class="tsd-kind-interface"><span class="tsd-kind-icon">Interface</span></li>
				<li class="tsd-kind-property tsd-parent-kind-interface"><span class="tsd-kind-icon">Property</span></li>
				<li class="tsd-kind-method tsd-parent-kind-interface"><span class="tsd-kind-icon">Method</span></li>
			</ul>
			<ul class="tsd-legend">
				<li class="tsd-kind-constructor tsd-parent-kind-class"><span class="tsd-kind-icon">Constructor</span></li>
				<li class="tsd-kind-property tsd-parent-kind-class"><span class="tsd-kind-icon">Property</span></li>
				<li class="tsd-kind-method tsd-parent-kind-class"><span class="tsd-kind-icon">Method</span></li>
			</ul>
			<ul class="tsd-legend">
				<li class="tsd-kind-method tsd-parent-kind-class tsd-is-inherited"><span class="tsd-kind-icon">Inherited method</span></li>
			</ul>
		</div>
	</div>
</footer>
<div class="container tsd-generator">
	<p>Generated using <a href="https://typedoc.org/" target="_blank">TypeDoc</a></p>
</div>
<div class="overlay"></div>
<script src="../assets/js/main.js"></script>
</body>
</html>