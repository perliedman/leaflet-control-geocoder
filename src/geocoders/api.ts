import * as L from 'leaflet';

/**
 * An object that represents a result from a geocoding query
 */
export interface GeocodingResult {
  /**
   * Name of found location
   */
  name: string;
  /**
   * The bounds of the location
   */
  bbox: L.LatLngBounds;
  /**
   * The center coordinate of the location
   */
  center: L.LatLng;
  /**
   * URL for icon representing result; optional
   */
  icon?: string;
  /**
   * HTML formatted representation of the name
   */
  html?: string;
  /**
   * Additional properties returned by the geocoder
   */
  properties?: any;
}

/**
 * A callback function used in {@link IGeocoder.geocode} and {@link IGeocoder.suggest} and {@link IGeocoder.reverse}
 */
export type GeocodingCallback = (result: GeocodingResult[]) => void;

/**
 * An interface implemented to respond to geocoding queries
 */
export interface IGeocoder {
  /**
   * Performs a geocoding query and returns the results to the callback in the provided context
   * @param query the query
   * @param cb the callback function
   * @param context the `this` context in the callback
   */
  geocode(query: string, cb: GeocodingCallback, context?: any): void;
  /**
   * Performs a geocoding query suggestion (this happens while typing) and returns the results to the callback in the provided context
   * @param query the query
   * @param cb the callback function
   * @param context the `this` context in the callback
   */
  suggest?(query: string, cb: GeocodingCallback, context?: any): void;
  /**
   * Performs a reverse geocoding query and returns the results to the callback in the provided context
   * @param location the coordinate to reverse geocode
   * @param scale the map scale possibly used for reverse geocoding
   * @param cb the callback function
   * @param context the `this` context in the callback
   */
  reverse?(location: L.LatLngLiteral, scale: number, cb: GeocodingCallback, context?: any): void;
}

export interface GeocoderOptions {
  /**
   * URL of the service
   */
  serviceUrl: string;
  /**
   * Additional URL parameters (strings) that will be added to geocoding requests
   */
  geocodingQueryParams?: Record<string, unknown>;
  /**
   * Additional URL parameters (strings) that will be added to reverse geocoding requests
   */
  reverseQueryParams?: Record<string, unknown>;
  /**
   * API key to use this service
   */
  apiKey?: string;
}

/**
 * @internal
 */
export function geocodingParams(
  options: GeocoderOptions,
  params: Record<string, unknown>
): Record<string, unknown> {
  return L.Util.extend(params, options.geocodingQueryParams);
}

/**
 * @internal
 */
export function reverseParams(
  options: GeocoderOptions,
  params: Record<string, unknown>
): Record<string, unknown> {
  return L.Util.extend(params, options.reverseQueryParams);
}
