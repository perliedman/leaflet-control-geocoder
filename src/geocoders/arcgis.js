import L from 'leaflet';
import { getJSON } from '../util';

export default {
  class: L.Class.extend({
    options: {
      service_url: 'http://geocode.arcgis.com/arcgis/rest/services/World/GeocodeServer'
    },

    initialize: function(accessToken, options) {
      L.setOptions(this, options);
      this._accessToken = accessToken;
    },

    geocode: function(query, cb, context) {
      var params = {
        SingleLine: query,
        outFields: 'Addr_Type',
        forStorage: false,
        maxLocations: 10,
        f: 'json'
      };

      if (this._key && this._key.length) {
        params.token = this._key;
      }

      getJSON(this.options.service_url + '/findAddressCandidates', params, function(data) {
        var results = [],
          loc,
          latLng,
          latLngBounds;

        if (data.candidates && data.candidates.length) {
          for (var i = 0; i <= data.candidates.length - 1; i++) {
            loc = data.candidates[i];
            latLng = L.latLng(loc.location.y, loc.location.x);
            latLngBounds = L.latLngBounds(
              L.latLng(loc.extent.ymax, loc.extent.xmax),
              L.latLng(loc.extent.ymin, loc.extent.xmin)
            );
            results[i] = {
              name: loc.address,
              bbox: latLngBounds,
              center: latLng
            };
          }
        }

        cb.call(context, results);
      });
    },

    suggest: function(query, cb, context) {
      return this.geocode(query, cb, context);
    },

    reverse: function(location, scale, cb, context) {
      var params = {
        location: encodeURIComponent(location.lng) + ',' + encodeURIComponent(location.lat),
        distance: 100,
        f: 'json'
      };

      getJSON(this.options.service_url + '/reverseGeocode', params, function(data) {
        var result = [],
          loc;

        if (data && !data.error) {
          loc = L.latLng(data.location.y, data.location.x);
          result.push({
            name: data.address.Match_addr,
            center: loc,
            bounds: L.latLngBounds(loc, loc)
          });
        }

        cb.call(context, result);
      });
    }
  }),

  factory: function(accessToken, options) {
    return new L.Control.Geocoder.ArcGis(accessToken, options);
  }
};
