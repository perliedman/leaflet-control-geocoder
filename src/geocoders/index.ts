export * from './api';
export * from './arcgis';
export * from './bing';
export * from './google';
export * from './here';
export * from './latlng';
export * from './mapbox';
export * from './mapquest';
export * from './neutrino';
export * from './nominatim';
export * from './open-location-code';
export * from './opencage';
export * from './pelias';
export * from './photon';
export * from './what3words';
