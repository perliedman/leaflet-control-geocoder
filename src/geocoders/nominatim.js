import L from 'leaflet';
import { template, getJSON } from '../util';

export default {
  class: L.Class.extend({
    options: {
      serviceUrl: 'https://nominatim.openstreetmap.org/',
      geocodingQueryParams: {},
      reverseQueryParams: {},
      htmlTemplate: function(r) {
        var a = r.address,
          parts = [];
        if (a.road || a.building) {
          parts.push('{building} {road} {house_number}');
        }

        if (a.city || a.town || a.village || a.hamlet) {
          parts.push(
            '<span class="' +
              (parts.length > 0 ? 'leaflet-control-geocoder-address-detail' : '') +
              '">{postcode} {city} {town} {village} {hamlet}</span>'
          );
        }

        if (a.state || a.country) {
          parts.push(
            '<span class="' +
              (parts.length > 0 ? 'leaflet-control-geocoder-address-context' : '') +
              '">{state} {country}</span>'
          );
        }

        return template(parts.join('<br/>'), a, true);
      }
    },

    initialize: function(options) {
      L.Util.setOptions(this, options);
    },

    geocode: function(query, cb, context) {
      getJSON(
        this.options.serviceUrl + 'search',
        L.extend(
          {
            q: query,
            limit: 5,
            format: 'json',
            addressdetails: 1
          },
          this.options.geocodingQueryParams
        ),
        L.bind(function(data) {
          var results = [];
          for (var i = data.length - 1; i >= 0; i--) {
            var bbox = data[i].boundingbox;
            for (var j = 0; j < 4; j++) bbox[j] = parseFloat(bbox[j]);
            results[i] = {
              icon: data[i].icon,
              name: data[i].display_name,
              html: this.options.htmlTemplate ? this.options.htmlTemplate(data[i]) : undefined,
              bbox: L.latLngBounds([bbox[0], bbox[2]], [bbox[1], bbox[3]]),
              center: L.latLng(data[i].lat, data[i].lon),
              properties: data[i]
            };
          }
          cb.call(context, results);
        }, this)
      );
    },

    reverse: function(location, scale, cb, context) {
      getJSON(
        this.options.serviceUrl + 'reverse',
        L.extend(
          {
            lat: location.lat,
            lon: location.lng,
            zoom: Math.round(Math.log(scale / 256) / Math.log(2)),
            addressdetails: 1,
            format: 'json'
          },
          this.options.reverseQueryParams
        ),
        L.bind(function(data) {
          var result = [],
            loc;

          if (data && data.lat && data.lon) {
            loc = L.latLng(data.lat, data.lon);
            result.push({
              name: data.display_name,
              html: this.options.htmlTemplate ? this.options.htmlTemplate(data) : undefined,
              center: loc,
              bounds: L.latLngBounds(loc, loc),
              properties: data
            });
          }

          cb.call(context, result);
        }, this)
      );
    }
  }),

  factory: function(options) {
    return new L.Control.Geocoder.Nominatim(options);
  }
};
