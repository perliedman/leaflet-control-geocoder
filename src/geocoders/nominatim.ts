import * as L from 'leaflet';
import { template, getJSON } from '../util';
import { GeocoderAPI, GeocodingResult } from './interfaces';

export interface NominatimResult {
  place_id: number;
  licence: string;
  osm_type: string;
  osm_id: number;
  boundingbox: string[];
  lat: string;
  lon: string;
  display_name: string;
  class: string;
  type: string;
  importance: number;
  icon?: string;
  address: Address;
}

export interface Address {
  building?: string;
  city_district?: string;
  city?: string;
  country_code?: string;
  country?: string;
  county?: string;
  hamlet?: string;
  house_number?: string;
  neighbourhood?: string;
  postcode?: string;
  road?: string;
  state_district?: string;
  state?: string;
  suburb?: string;
  village?: string;
}

export interface NominatimOptions {
  serviceUrl: string;
  geocodingQueryParams?: object;
  reverseQueryParams?: object;
  htmlTemplate: (r: NominatimResult) => string;
}

export class Nominatim implements GeocoderAPI {
  options: NominatimOptions;

  constructor(options?: NominatimOptions) {
    this.options = L.Util.extend(
      {
        serviceUrl: 'https://nominatim.openstreetmap.org/',
        geocodingQueryParams: {},
        reverseQueryParams: {},
        htmlTemplate: function(r: NominatimResult) {
          var a = r.address,
            className,
            parts = [];
          if (a.road || a.building) {
            parts.push('{building} {road} {house_number}');
          }

          if (a.city || (a as any).town || a.village || a.hamlet) {
            className = parts.length > 0 ? 'leaflet-control-geocoder-address-detail' : '';
            parts.push(
              '<span class="' + className + '">{postcode} {city} {town} {village} {hamlet}</span>'
            );
          }

          if (a.state || a.country) {
            className = parts.length > 0 ? 'leaflet-control-geocoder-address-context' : '';
            parts.push('<span class="' + className + '">{state} {country}</span>');
          }

          return template(parts.join('<br/>'), a);
        }
      },
      options
    );
  }

  geocode(query: string, cb: (result: GeocodingResult[]) => void, context?: any) {
    getJSON(
      this.options.serviceUrl + 'search',
      L.Util.extend(
        {
          q: query,
          limit: 5,
          format: 'json',
          addressdetails: 1
        },
        this.options.geocodingQueryParams
      ),
      data => {
        var results = [] as GeocodingResult[];
        for (var i = data.length - 1; i >= 0; i--) {
          var bbox = data[i].boundingbox;
          for (var j = 0; j < 4; j++) bbox[j] = parseFloat(bbox[j]);
          results[i] = {
            icon: data[i].icon,
            name: data[i].display_name,
            html: this.options.htmlTemplate ? this.options.htmlTemplate(data[i]) : undefined,
            bbox: L.latLngBounds([bbox[0], bbox[2]], [bbox[1], bbox[3]]),
            center: L.latLng(data[i].lat, data[i].lon),
            properties: data[i]
          };
        }
        cb.call(context, results);
      }
    );
  }

  reverse(location: L.LatLng, scale: number, cb: (result: any) => void, context?: any) {
    getJSON(
      this.options.serviceUrl + 'reverse',
      L.Util.extend(
        {
          lat: location.lat,
          lon: location.lng,
          zoom: Math.round(Math.log(scale / 256) / Math.log(2)),
          addressdetails: 1,
          format: 'json'
        },
        this.options.reverseQueryParams
      ),
      data => {
        var result = [],
          loc;

        if (data && data.lat && data.lon) {
          loc = L.latLng(data.lat, data.lon);
          result.push({
            name: data.display_name,
            html: this.options.htmlTemplate ? this.options.htmlTemplate(data) : undefined,
            center: loc,
            bounds: L.latLngBounds(loc, loc),
            properties: data
          });
        }
        cb.call(context, result);
      }
    );
  }
}

export function nominatim(options: NominatimOptions) {
  return new Nominatim(options);
}
