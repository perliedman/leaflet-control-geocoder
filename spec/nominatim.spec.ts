import { testXMLHttpRequest } from './mockXMLHttpRequest';
import { Nominatim } from '../src/geocoders/nominatim';

describe('L.Control.Geocoder.Nominatim', () => {
  const geocoder = new Nominatim();

  it('geocodes Innsbruck', () => {
    const callback = jest.fn();

    testXMLHttpRequest(
      'https://nominatim.openstreetmap.org/search?q=innsbruck&limit=5&format=json&addressdetails=1',
      [
        {
          place_id: 199282228,
          licence: 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright',
          osm_type: 'relation',
          osm_id: 8182617,
          boundingbox: ['47.2583715', '47.2808566', '11.3811871', '11.418183'],
          lat: '47.26951525',
          lon: '11.3971372042211',
          display_name: 'Innsbruck, Tyrol, Austria',
          class: 'boundary',
          type: 'administrative',
          importance: 0.763909048330467,
          icon:
            'https://nominatim.openstreetmap.org/images/mapicons/poi_boundary_administrative.p.20.png',
          address: {
            city_district: 'Innsbruck',
            city: 'Innsbruck',
            county: 'Innsbruck',
            state: 'Tyrol',
            country: 'Austria',
            country_code: 'at'
          }
        }
      ],
      () => geocoder.geocode('innsbruck', callback)
    );

    const feature = callback.mock.calls[0][0][0];
    expect(feature.name).toBe('Innsbruck, Tyrol, Austria');
    expect(feature.html).toBe(
      '<span class=""> Innsbruck   </span><br/><span class="leaflet-control-geocoder-address-context">Tyrol Austria</span>'
    );
    expect(feature.properties.address).toStrictEqual({
      city_district: 'Innsbruck',
      city: 'Innsbruck',
      county: 'Innsbruck',
      state: 'Tyrol',
      country: 'Austria',
      country_code: 'at'
    });
    expect(callback.mock.calls).toMatchSnapshot();
  });

  it('reverse geocodes 47.3/11.3', () => {
    const callback = jest.fn();

    testXMLHttpRequest(
      'https://nominatim.openstreetmap.org/reverse?lat=47.3&lon=11.3&zoom=9&addressdetails=1&format=json',
      {
        place_id: 197718025,
        licence: 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright',
        osm_type: 'relation',
        osm_id: 78251,
        lat: '47.2065094',
        lon: '11.3836945900354',
        display_name: 'Innsbruck-Land, Tyrol, Austria',
        address: {
          county: 'Innsbruck-Land',
          state: 'Tyrol',
          country: 'Austria',
          country_code: 'at'
        },
        boundingbox: ['46.9624854', '47.4499229', '10.9896868', '11.7051742']
      },
      () => geocoder.reverse({ lat: 47.3, lng: 11.3 }, 131000, callback)
    );

    const feature = callback.mock.calls[0][0][0];
    expect(feature.name).toBe('Innsbruck-Land, Tyrol, Austria');
    expect(feature.html).toBe('<span class="">Tyrol Austria</span>');
    expect(feature.properties.address).toStrictEqual({
      county: 'Innsbruck-Land',
      state: 'Tyrol',
      country: 'Austria',
      country_code: 'at'
    });
    expect(callback.mock.calls).toMatchSnapshot();
  });
});
