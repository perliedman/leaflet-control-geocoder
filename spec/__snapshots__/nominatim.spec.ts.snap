// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`L.Control.Geocoder.Nominatim geocodes Innsbruck 1`] = `
Array [
  Array [
    Array [
      Object {
        "bbox": Object {
          "_northEast": Object {
            "lat": 47.2808566,
            "lng": 11.418183,
          },
          "_southWest": Object {
            "lat": 47.2583715,
            "lng": 11.3811871,
          },
        },
        "center": Object {
          "lat": 47.26951525,
          "lng": 11.3971372042211,
        },
        "html": "<span class=\\"\\"> Innsbruck   </span><br/><span class=\\"leaflet-control-geocoder-address-context\\">Tyrol Austria</span>",
        "icon": "https://nominatim.openstreetmap.org/images/mapicons/poi_boundary_administrative.p.20.png",
        "name": "Innsbruck, Tyrol, Austria",
        "properties": Object {
          "address": Object {
            "city": "Innsbruck",
            "city_district": "Innsbruck",
            "country": "Austria",
            "country_code": "at",
            "county": "Innsbruck",
            "state": "Tyrol",
          },
          "boundingbox": Array [
            47.2583715,
            47.2808566,
            11.3811871,
            11.418183,
          ],
          "class": "boundary",
          "display_name": "Innsbruck, Tyrol, Austria",
          "icon": "https://nominatim.openstreetmap.org/images/mapicons/poi_boundary_administrative.p.20.png",
          "importance": 0.763909048330467,
          "lat": "47.26951525",
          "licence": "Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright",
          "lon": "11.3971372042211",
          "osm_id": 8182617,
          "osm_type": "relation",
          "place_id": 199282228,
          "type": "administrative",
        },
      },
    ],
  ],
]
`;

exports[`L.Control.Geocoder.Nominatim reverse geocodes 47.3/11.3 1`] = `
Array [
  Array [
    Array [
      Object {
        "bbox": Object {
          "_northEast": Object {
            "lat": 47.2065094,
            "lng": 11.3836945900354,
          },
          "_southWest": Object {
            "lat": 47.2065094,
            "lng": 11.3836945900354,
          },
        },
        "center": Object {
          "lat": 47.2065094,
          "lng": 11.3836945900354,
        },
        "html": "<span class=\\"\\">Tyrol Austria</span>",
        "name": "Innsbruck-Land, Tyrol, Austria",
        "properties": Object {
          "address": Object {
            "country": "Austria",
            "country_code": "at",
            "county": "Innsbruck-Land",
            "state": "Tyrol",
          },
          "boundingbox": Array [
            "46.9624854",
            "47.4499229",
            "10.9896868",
            "11.7051742",
          ],
          "display_name": "Innsbruck-Land, Tyrol, Austria",
          "lat": "47.2065094",
          "licence": "Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright",
          "lon": "11.3836945900354",
          "osm_id": 78251,
          "osm_type": "relation",
          "place_id": 197718025,
        },
      },
    ],
  ],
]
`;
